### PR TITLE
feat(pipeline): support common workflow

### PR DESCRIPTION
**Pull request by facethesunsea**
### What this PR does / Why we need it:

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

### Check List 

- [ ] Docs have been added / updated
- [ ] Unit test / Integration test for the changes have been added
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

## More information

### CODE REVIEW

File: src/components/projects/edit_pipeline/common_pipeline/pipeline.vue [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+          this.selectedTabs.forEach(tab => {
+            if (tab.type) {
+              commonInfo.sub_tasks.push(commonInfo[tab.type])
+            }
+          })
+
+          delete commonInfo.buildv3
+          delete commonInfo.trigger
+
+          console.log('保存', commonInfo)
```
Comment (leozhang2018) [inline src/components/projects/edit_pipeline/common_pipeline/pipeline.vue:107]: Remove console.log

**Review by leozhang2018 — CHANGES_REQUESTED** [review]

File: src/components/projects/edit_pipeline/common_pipeline/switch_tab/basic_dialog.vue [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+import { cloneDeep } from 'lodash'
+
+const paramInfo = {
+  key: '',
+  type: '',
+  default_value: '',
+  choice_option: '',
+  external_setting: {
+    system_id: '',
+    // Endpoint路径
```
Comment (leozhang2018) [inline src/components/projects/edit_pipeline/common_pipeline/switch_tab/basic_dialog.vue:95]: Please use English to comment

File: src/components/projects/edit_pipeline/common_pipeline/switch_tab/basic_info.vue [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    },
+    summit () {
+      const res = this.$refs.paramRef.getParamData()
+      res.choice_option = res.choice_option.split(',')
+      this.commonInfoUse.parameters.splice(this.paramIndex, 1, res)
+      this.dialogVisible = false
+    },
+    validate () {
+      return this.$refs.basicForm.validate()
+      // return Promise.resolve()
```
Comment (leozhang2018) [inline src/components/projects/edit_pipeline/common_pipeline/switch_tab/basic_info.vue:143]: Remove it or add some comment.
For example:
  return this.$refs.basicForm.validate()
  //Todo:use promise to check in furture.
      // return Promise.resolve()

File: src/components/projects/edit_pipeline/common_pipeline/switch_tab/build.vue [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+                <el-option label="高 | CPU: 16 核 内存: 32 GB" value="high"></el-option>
+                <el-option label="中 | CPU: 8 核 内存: 16 GB" value="medium"></el-option>
+                <el-option label="低 | CPU: 4 核 内存: 8 GB" value="low"></el-option>
+                <el-option label="最低 | CPU: 2 核 内存: 2 GB" value="min"></el-option>
+              </el-select>
+            </el-form-item>
+          </el-col>
+        </el-row>
+      </el-form>
+      <el-form ref="buildApp" :inline="true" :model="buildInfo" class="form-style1" label-position="top" label-width="80px">
```
Comment (leozhang2018) [inline src/components/projects/edit_pipeline/common_pipeline/switch_tab/build.vue:35]: class="form-style1" 
Please use meaningful name

File: src/components/projects/pipeline/common/run_common_workflow.vue [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+          console.log(res)
+
+          const buildStep = res.sub_tasks.find(task => task.type === 'buildv3')
+          // 没有构建不能运行工作流
+          if (!buildStep) {
+            this.cantRun = true
+            return
+          }
+
+          const infos = [] // 用来请求branch pr信息
```
Comment (leozhang2018) [inline src/components/projects/pipeline/common/run_common_workflow.vue:102]: Please use English.

**Review by leozhang2018 — CHANGES_REQUESTED** [review]

File: src/components/projects/edit_pipeline/common_pipeline/switch_tab/build.vue (hunk 4 above)
Comment (facethesunsea) [reply src/components/projects/edit_pipeline/common_pipeline/switch_tab/build.vue:35]: all build use the class name, update all later I think.

File: src/components/projects/pipeline/workflow_list.vue [hunk 6]
```diff
@@ -234,10 +268,16 @@ export default {
           return []
         })
       }
-      if (res) {
-        this.workflowsList = res
-        this.workflowListLoading = false
-      }
+
+      const res2 = await getCommonWorkflowListAPI(projectName).catch(err => {
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list.vue:272]: Please use meaningful variable name.

File: src/components/projects/edit_pipeline/common_pipeline/pipeline.vue [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+      <el-card class="content">
+        <keep-alive>
+          <component :is="isComp" :validObj="validObj" ref="commonStep" :isEdit="!!pipelineId"></component>
+        </keep-alive>
+      </el-card>
+    </div>
+    <footer class="footer">
+      <el-tag>填写相关信息，然后点击保存</el-tag>
+      <el-button type="primary" plain @click="$router.go(-1)">取消</el-button>
+      <el-button type="primary" plain @click="saveCommonPipeline">保存</el-button>
```
Comment (leozhang2018) [inline src/components/projects/edit_pipeline/common_pipeline/pipeline.vue:15]: CommonPipeline -> CommonWorkflow

**Review by leozhang2018 — CHANGES_REQUESTED** [review]

File: src/store/mutations.js [hunk 8]
```diff
@@ -36,3 +36,10 @@ export const QUERY_ORDER_SERVICE = 'QUERY_ORDER_SERVICE'
 
 // External Link
 export const SET_EXTERNAL_LINK = 'SET_EXTERNAL_LINK'
+
+// common pipeline
```
Comment (leozhang2018) [inline src/store/mutations.js:40]: //Common Workflow

File: src/utilities/validate_async.js [hunk 9]
```diff
@@ -1,6 +1,6 @@
 export default class ValidateSubmit {
   constructor () {
-    this.validateArr = []
+    this.validateArr = [] // []{name, valid} name唯一
```
Comment (leozhang2018) [inline src/utilities/validate_async.js:3]: Please use English

File: src/components/projects/pipeline/workflow_list.vue [hunk 10]
```diff
@@ -234,10 +268,16 @@ export default {
...
+        console.log(err)
+        return []
+      })
+      res2.workflow_list.forEach(list => {
+        list.type = 'common'
+      })
+      this.workflowListLoading = false
+      this.workflowsList = [...res, ...res2.workflow_list]
     },
     deleteWorkflow (workflow) {
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list.vue:282]: deleteProductWorkflow

File: src/components/projects/pipeline/workflow_list.vue [hunk 11]
```diff
@@ -262,7 +302,31 @@ export default {
         })
       })
     },
+    deleteCommon (workflow) {
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list.vue:305]: deleteCommonWorkflow

File: src/components/projects/pipeline/workflow_list.vue [hunk 12]
```diff
@@ -313,16 +377,22 @@ export default {
...
-        this.$router.push(`/workflows/edit/${newName}?projectName=${projectName}`)
+        this.$router.push(
+          `/workflows/product/edit/${newName}?projectName=${projectName}`
+        )
       })
     },
     sortWorkflow (cm) {
       this.sortBy = cm
+    },
+    startCommonBuild (worflow) {
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list.vue:388]: startCommonWorkflowBuild

File: src/components/projects/pipeline/workflow_list.vue [hunk 13]
```diff
@@ -262,7 +302,31 @@ export default {
...
+          deleteCommonWorkflowAPI(workflow.project_name, workflow.id).then(res => {
+            this.getWorkflows(this.projectName)
+            this.$message.success(`${value}删除成功！`)
+          })
+        })
+        .catch(() => {
+          this.$message.info('取消删除')
+        })
+    },
     startProductBuild (workflow) {
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list.vue:328]: startProductWorkflowBuild

File: src/components/projects/pipeline/workflow_list/common_row.vue [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+      favorite: false,
+      testId: '1'
+    }
+  },
+  props: {
+    workflow: Object
+  },
+  inject: ['startCommonBuild', 'deleteCommon'],
+  methods: {
+    deleteCommonPipeline (command) {
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list/common_row.vue:77]: deleteCommonWorkflow

**Review by leozhang2018 — CHANGES_REQUESTED** [review]

File: src/components/projects/edit_pipeline/common_pipeline/switch_tab/basic_dialog.vue [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+    endpoint: '',
+    // 请求方法
```
Comment (leozhang2018) [inline src/components/projects/edit_pipeline/common_pipeline/switch_tab/basic_dialog.vue]: Please use English

File: src/components/projects/edit_pipeline/common_pipeline/switch_tab/external.vue [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+      },
+      path: {
+        required: true,
+        message: '请输入访问路径',
+        trigger: ['blur', 'change']
+      }
+    }
+    return {
+      externalData: cloneDeep(externalInfo),
+      externalList: [
```
Comment (leozhang2018) [inline src/components/projects/edit_pipeline/common_pipeline/switch_tab/external.vue:82]: Please use empty array.

**Review by leozhang2018 — CHANGES_REQUESTED** [review]

File: src/components/projects/pipeline/workflow_list/virtual_list_item.vue [hunk 17]
```diff
@@ -1,6 +1,7 @@
 <template>
   <div>
-    <pipeline-row :name="workflow.name"
+    <CommonRow v-if="source.type === 'common'" :workflow="source"></CommonRow>
+    <pipeline-row v-else :name="workflow.name"
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list/virtual_list_item.vue:4]: ProductWorkflowRow

File: src/components/projects/pipeline/workflow_list/virtual_list_item.vue [hunk 18]
```diff
@@ -1,6 +1,7 @@
 <template>
   <div>
-    <pipeline-row :name="workflow.name"
+    <CommonRow v-if="source.type === 'common'" :workflow="source"></CommonRow>
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list/virtual_list_item.vue:3]: CommonWorkflowRow

File: src/components/projects/pipeline/workflow_list/virtual_list_item.vue [hunk 19]
```diff
@@ -55,6 +56,7 @@
 
 <script>
 import pipelineRow from './pipeline_row.vue'
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list/virtual_list_item.vue:58]: ProductWorkflowRow

File: src/components/projects/pipeline/workflow_list/virtual_list_item.vue [hunk 20]
```diff
@@ -55,6 +56,7 @@
...
+import CommonRow from './common_row.vue'
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list/virtual_list_item.vue:59]: CommonWorkflowRow

File: src/components/setting/integration/other.vue [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+      this.dialogExternalVisible = true
+      this.externalEdit = cloneDeep(row)
+      getExternalSystemByIdAPI(row.id).then(res => {
+        if (this.externalEdit.id === res.id) {
+          this.externalEdit = res
+        }
+      })
+    },
+    handleExternalDelete (id) {
+      this.$confirm(`确定要删除这个 External 配置吗？`, '确认', {
```
Comment (leozhang2018) [inline src/components/setting/integration/other.vue:134]: External 配置 ->外部系统配置

**Review by leozhang2018 — CHANGES_REQUESTED** [review]

File: src/components/projects/pipeline/common/run_common_workflow.vue [hunk 22]
```diff
@@ -0,0 +1,247 @@
...
+          const key = `${re.codehost_id}/${re.repo_owner}/${re.repo}`
+          const build = builds.find(build => build.key === key)
+          if (build) {
+            build.branches = re.branches
+            build.prs = re.prs
+          }
+        })
+      })
+    },
+    getCommonPipelineInfo () {
```
Comment (leozhang2018) [inline src/components/projects/pipeline/common/run_common_workflow.vue:92]: getCommonWorkflowInfo

File: src/components/projects/pipeline/workflow_list.vue [hunk 23]
```diff
@@ -59,24 +56,42 @@
...
+        <div class="type-desc">可自定义工作流程，内置构建、K8s 部署、小程序发版等步骤</div>
+      </div>
+      <div slot="footer">
+        <el-button size="small" @click="showSelectWorkflowType = false">取 消</el-button>
+        <el-button size="small" type="primary" @click="createWorkflow">确 定</el-button>
+      </div>
+    </el-dialog>
+
     <el-dialog title="运行 产品-工作流" :visible.sync="showStartProductBuild" custom-class="run-workflow" width="60%">
       <run-workflow
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list.vue:73]: RunProductWorkflow

File: src/components/projects/pipeline/workflow_list.vue [hunk 24]
```diff
@@ -59,24 +56,42 @@
...
+
+    <el-dialog title="运行 通用-工作流" :visible.sync="showStartCommonWorkflowBuild" :close-on-click-modal="false">
+      <RunCommonWorkflow :value="showStartCommonWorkflowBuild" :workflow="commonToRun"></RunCommonWorkflow>
+    </el-dialog>
   </div>
 </template>
 
 <script>
 import VirtualListItem from './workflow_list/virtual_list_item'
 import runWorkflow from './common/run_workflow.vue'
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list.vue:90]: RunProductWorkflow

File: src/components/projects/pipeline/workflow_list.vue [hunk 25]
```diff
@@ -313,16 +377,22 @@ export default {
...
-        this.$router.push(`/workflows/edit/${newName}?projectName=${projectName}`)
+        this.$router.push(
+          `/workflows/product/edit/${newName}?projectName=${projectName}`
+        )
       })
     },
     sortWorkflow (cm) {
       this.sortBy = cm
+    },
+    StartCommonWorkflowBuild (worflow) {
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list.vue:388]: StartCommonWorkflowBuild -> startCommonWorkflowBuild

File: src/components/projects/pipeline/workflow_list/common_row.vue [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+  data () {
+    return {
+      favorite: false,
+      testId: '1'
+    }
+  },
+  props: {
+    workflow: Object
+  },
+  inject: ['StartCommonWorkflowBuild', 'deleteCommonWorkflow']
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_list/common_row.vue:75]: startCommonWorkflowBuild -> StartCommonWorkflowBuild

**Review by leozhang2018 — CHANGES_REQUESTED** [review]

File: src/components/projects/pipeline/common/run_common_workflow.vue [hunk 27]
```diff
@@ -0,0 +1,247 @@
...
+            </el-select>
+            <el-select v-else-if="row.type === 'external'" v-model="row.value" placeholder="请选择值" size="small" :value-key="row.key">
+              <el-option v-for="val in row.options" :key="val[row.key]" :label="stringFromData(val[row.key])" :value="val"></el-option>
+            </el-select>
+          </template>
+        </el-table-column>
+      </el-table>
+    </section>
+    <footer>
+      <el-button :loading="loading" type="primary" plain size="small" :disabled="cantRun" @click="runCommonPipeline">启动任务</el-button>
```
Comment (leozhang2018) [inline src/components/projects/pipeline/common/run_common_workflow.vue:46]: runCommonWorkflow

File: src/components/projects/pipeline/workflow_common_detail.vue [hunk 28]
```diff
@@ -0,0 +1,494 @@
...
+      const res = await getCommonWorkflowTasksAPI(
+        projectName,
+        workflowName,
+        pageStart,
+        pageSize
+      )
+      this.workflowTasks = res.data
+      this.total = res.total
+      if (!this.timeTimeoutFinishFlag) {
+        this.timerId = setTimeout(this.autoRefreshHistoryTask, 3000) // just one timeout
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_common_detail.vue:151]: Keep only one timer

File: src/components/projects/pipeline/workflow_common_detail.vue [hunk 29]
```diff
@@ -0,0 +1,494 @@
...
+      )
+      this.workflowTasks = res.data
+      this.total = res.total
+    },
+    changeTaskPage (val) {
+      const start = (val - 1) * this.pageSize
+      this.pageStart = start
+      this.getCommonWorkflowTasks(start, this.pageSize)
+    },
+    hideAndgetCommonWorkflowTasks () {
```
Comment (leozhang2018) [inline src/components/projects/pipeline/workflow_common_detail.vue:171]: hideAndGetCommonWorkflowTasks

**Review by leozhang2018 — CHANGES_REQUESTED** [review]

File: src/store/index.js [hunk 30]
```diff
@@ -15,6 +15,9 @@ import service_manage from './modules/service_manage'
 // External Link
 import external_link from './modules/external_link'
 
+// common workflow
```
Comment (leozhang2018) [inline src/store/index.js:18]: Common Workflow

File: src/store/index.js [hunk 31]
```diff
@@ -15,6 +15,9 @@ import service_manage from './modules/service_manage'
...
+import common_pipeline from './modules/common_pipeline'
```
Comment (leozhang2018) [inline src/store/index.js:19]: common_workflow

**Review by leozhang2018 — CHANGES_REQUESTED** [review]

**Review by leozhang2018 — APPROVED** [review]
LGTM

**Pull request merged**